remotefiles.go

```diff
@@ -0,0 +1,286 @@
+	Channels        []string `json:"channels"`
+	Groups          []string `json:"groups"`
+	IMs             []string `json:"ims"`
+	ExternalID      string   `json:"external_id"`
+	ExternalURL     string   `json:"external_url"`
+	HasRichPreview  bool     `json:"has_rich_preview"`
+}
+
+type RemoteFileParameters struct {
+	ExternalID            string // required
+	ExternalURL           string // required
+	Title                 string // required
+	Filetype              string
+	IndexableFileContents string
+	PreviewImage          string
+	PreviewImageReader    io.Reader
+}
+
+type ListRemoteFilesParameters struct {
+	Channel       string
+	Cursor        string
+	Limit         int
+	TimestampFrom JSONTime
+	TimestampTo   JSONTime
+}
+
+type remoteFileResponseFull struct {
+	RemoteFile `json:"file"`
+	Paging     `json:"paging"`
+	Files      []RemoteFile `json:"files"`
+	SlackResponse
+}
+
+func (api *Client) remoteFileRequest(ctx context.Context, path string, values url.Values) (*remoteFileResponseFull, error) {
+	response := &remoteFileResponseFull{}
+	err := api.postMethod(ctx, path, values, response)
+	if err != nil {
+		return nil, err
+	}
+
+	return response, response.Err()
+}
+
+// AddRemoteFile adds a remote file
+func (api *Client) AddRemoteFile(params RemoteFileParameters) (remotefile *RemoteFile, err error) {
+	return api.AddRemoteFileContext(context.Background(), params)
+}
+
+// AddRemoteFileContext adds a remote file and setting a custom context
+func (api *Client) AddRemoteFileContext(ctx context.Context, params RemoteFileParameters) (remotefile *RemoteFile, err error) {
+	if params.ExternalID == "" || params.ExternalURL == "" || params.Title == "" {
+		return nil, ErrParametersMissing
+	}
+	response := &remoteFileResponseFull{}
+	values := url.Values{
+		"token":        {api.token},
+		"external_id":  {params.ExternalID},
+		"external_url": {params.ExternalURL},
+		"title":        {params.Title},
+	}
+	if params.Filetype != "" {
+		values.Add("filetype", params.Filetype)
+	}
+	if params.IndexableFileContents != "" {
+		values.Add("indexable_file_contents", params.IndexableFileContents)
+	}
+	if params.PreviewImage != "" {
+		err = postLocalWithMultipartResponse(ctx, api.httpclient, api.endpoint+"files.remote.add", params.PreviewImage, "preview_image", values, response, api)
+	} else if params.PreviewImageReader != nil {
+		err = postWithMultipartResponse(ctx, api.httpclient, api.endpoint+"files.remote.add", "preview.png", "preview_image", values, params.PreviewImageReader, response, api)
+	} else {
+		response, err = api.remoteFileRequest(ctx, "files.remote.add", values)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &response.RemoteFile, response.Err()
+}
+
+// ListRemoteFiles retrieves all remote files according to the parameters given. Uses cursor based pagination.
+func (api *Client) ListRemoteFiles(params ListRemoteFilesParameters) ([]RemoteFile, *ListRemoteFilesParameters, error) {
```

Comment:
How about using string instead of `ListRemoteFilesParameters` in returned value?
I think `ListRemoteFilesParameters` should be used only as an argument to ListRemoteFiles.